Explainer rewrite (#179)
WICG/shared-element-transitions: explainer.md

# Introduction

https://user-images.githubusercontent.com/93594/184085118-65b33a92-272a-49f4-b3d6-50a8b8313567.mp4

> Page transitions not only look great, they also communicate direction of flow, and make it clear which elements are related from page to page. They can even happen during data fetching, leading to a faster perception of performance.

Author: I removed it because the "flash of white" is wrong, and it seemed to focus on MPA. But yeah, cutting it has lost too much. I'll add some of this back.

On Single Page Apps (SPAs) developers can use animation tools such as [CSS transitions](https://developer.mozilla.org/docs/Web/CSS/CSS_Transitions/Using_CSS_transitions), [CSS animations](https://developer.mozilla.org/docs/Web/CSS/CSS_Animations/Using_CSS_animations), and the [Web Animation API](https://developer.mozilla.org/docs/Web/API/Web_Animations_API/Using_the_Web_Animations_API) to create transitions. However, it's not straight forward, mostly due to the period where both states need to exist in the DOM at the same time, so both can be partially visible (eg in a cross-fade). This creates issues with:

- **Accessibility**: Having both states in the DOM at the same time can create a confusing experience for screen reader users. Juggling the DOM around for the sake of a transition doesn't play well with things like ARIA live regions.

> - **Usability**: The old DOM will continue to exist during the transition-away, leading to complications if the user manages to interact with that content (eg clicking buttons).

Reviewer: ```suggestion
- **Usability**: The old DOM will continue to exist during the transition, leading to complications if the user manages to interact with that content (eg clicking buttons).
```

The `prepare()` callback allows for the DOM change to be async, as many frameworks batch "please change this state" requests. Due to the render blocking during this phase, browsers will have an aggressive timeout. Developers shouldn't use this to block rendering while fetching view data - this should be done before calling `prepare()`.

## The captured state

Assuming a page's header, and the text within that header, are to be animated independently:

```css
.header {
  page-transition-tag: header;
  contain: layout;
}
.header-text {
  page-transition-tag: header-text;
  contain: layout;
}
```

The page will be captured in three parts: The header, the header text, and the remaining page (known as the 'root').

https://user-images.githubusercontent.com/93594/184097864-40b9c860-480a-45ff-9787-62cebe68a078.mp4

_The above shows the parts of the page captured before and after the DOM change._

For the outgoing state, the following is captured for each part:

> - An image of the element, including ink overflow if any. The browser may crop this to make efficient use of memory, and that cropping should be biased towards the viewport.

Reviewer: ```suggestion
- An image of the element, including ink overflow if any. The browser may crop this to make efficient use of memory, and that cropping should be biased towards the viewport. This image also excludes painting of any descendants with a page-transition-tag, as if they have `visibility: hidden`.
```

- The layout dimensions.
- A CSS transform that would place element from the layout viewport origin to its current quad.
- An [`object-view-box`](https://drafts.csswg.org/css-images-4/#propdef-object-view-box) that coincides with the element's border box.

> The above is also captured for the incoming state, although the image of the element is 'live', as in it's connected to the representation currently in the DOM.

Reviewer: ```suggestion
The above is also captured for the incoming state, although the image and computed layout properties of the element are 'live', as in it's connected to the representation currently in the DOM.
```

It's valid for some transition elements to only exist on one side of the DOM change.

## Building the pseudo-element tree

For each transitioning element (including the root), the following pseudo-element tree is created:

```
::page-transition-container(name)
└─ ::page-transition-image-wrapper(name)
   ├─ ::page-transition-outgoing-image(name)
   └─ ::page-transition-incoming-image(name)
```

> Where `name` is the `page-transition-tag` value, which is `root` for the root. If the transitioning element only existed on one side of the DOM change, then either the outgoing or incoming image will be missing.

Reviewer: Should we mention that tagging the root element is configurable. Its implicitly captured using the following UA style rule:

```css
html {
  page-transition-tag: root;
}
```

But developers can rename it or reset it to none.

Author: I'll add a note

> These trees are inserted into a `::page-transition` pseudo element, according to the render-order of their transition elements.

Reviewer: ```suggestion
These trees are inserted into a `::page-transition` pseudo element, according to the paint order of their transition elements.
```

Its a nit since we've used the word paint order in all other documentation.

> The `::page-transition` is rendered in a top-level, filling the viewport.

Reviewer: ```suggestion
The `::page-transition` is rendered in a top-level stacking context, filling the viewport.
```

I assumed you meant to write something after top-level, not sure if this was it.

Once this is in place, rendering is resumed.

# Default styles & animation

## `::page-transition-container(*)`

> Default styles:

Reviewer: This section would read easier if we just put the style declaration instead of the text. Example for the container:

```css
html::page-transition-container(foo) {
  position: absolute;

  /* Width and height of the incoming element, if any, otherwise the outgoing element. */
  width: 100px;
  height: 100px;

  /* A transform that places it in the viewport position of the incoming element, if any, otherwise the outgoing element */
  transform: translate(100px, 0);
}
```

Author: Agreed. I avoided it because I wasn't sure if it'd be clear the styles were generated rather than static.

- Absolutely positioned to the top left of the parent.

> - Width and height of the outgoing element.

Reviewer: Since as the spec we create the pseudo DOM only when both the incoming and outgoing state are ready, all the computed properties correspond to the incoming elements if we have any.

Author: Oh, so it starts with the incoming size, and animates from the outgoing size?

Reviewer: Yup. The width/height corresponds to the to state and the keyframe provides the from state. Internally its the outgoing size until we hit the running phase but since this is web exposed when we get to the running/animation phase, at that point all values are from the incoming one.

- A transform that places it in the viewport position of the outgoing element.

Default animation:

> - Width and height animates to the dimensions of the incoming element.

Reviewer: Here too, the precise CSS animation being created by the browser makes it clearer. One of the confusion that Tab had during the spec review was whether we were using WebAnimations for the default animations.

Author: We can chat about this later, as it's not 100% clear to me

## `::page-transition-image-wrapper(*)`

Default styles:

- Absolutely positioned with 0 inset.
- `isolation: isolate` to aid with cross-fading.

Default animation: none.

## `::page-transition-outgoing-image(*)`

Default styles:

- A replaced element displaying the capture of the outgoing element, with a natural aspect ratio of the outgoing element.
- Absolutely positioned to the inline & block start.
- `mix-blend-mode: plus-lighter` to allow for a true cross-fade.
- 100% block size
- Auto inline size
- The captured `object-view-box`

The `object-view-box` allows the image to be the layout size of the element, but allow overflow (to accommodate ink-overflow) and underflow (cropping to save memory) in the image data.

Default animation:

- Opacity animates from 1 to 0.

Note that the height of this element is auto, so it won't stretch the image as the container changes height. The developer can change this if they wish.

## `::page-transition-incoming-image(*)`

As above, but animates from opacity 0 to 1.

# Ending the transition

The transition ends once a state is reached where no animations are running on any of the transition pseuduo-elements.

> Once this happens, the pseudo-elements are removed, and the real DOM underneath is shown.

Reviewer: We should clarify somewhere before this that the tagged DOM elements are not painted in the real DOM while they are being displayed via pseudo-elements during the transition. The effect is similar to the elements and their 
descendants having `visibility: hidden` and `pointer-events: none`.

Due to the simple nature of these pseudo-element trees, these animations should be able to run off the main thread. However, if the developer adds something that requires layout, such as a border, the animation will fall back to main thread.

# Compatibility with frameworks

```js
const transition = new SameDocumentTransition();
await transition.prepare(async () => {
  await updateTheDOMSomehow();
});
```

This pattern assumes the developer is in charge of DOM updates, but that isn't the case with most web frameworks. However, the [demo site featured in this explainer](https://http203-playlist.netlify.app/) was built using Preact, and uses a [React-style hook](https://github.com/jakearchibald/http203-playlist/blob/main/src/client/utils.ts#L11) to wrap the above API and make it usable with React/Preact.

As long as the framework provides a notification when the DOM is updated, which they already do to allow custom handling of elements, the transition API can be made to work with the framework.

# Animating with JavaScript

> The promise returned by `prepare()` fulfills when both states have been captures and the pseudo-element tree has been successfully built. This provides developers with a point where they can animate those pseudo-elements with the [Web Animation API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API).

Reviewer: ```suggestion
The promise returned by `prepare()` fulfills when both states have been captured and the pseudo-element tree has been successfully built. This provides developers with a point where they can animate those pseudo-elements with the [Web Animation API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API).
```

https://user-images.githubusercontent.com/93594/184120371-678f58b3-d1f9-465b-978f-ee5eab73d120.mp4

This example animates a circular `clip-path` from the position of the click to a size that covers the viewport. Since this isn't possible with CSS alone ([although it may get easier in future](https://github.com/w3c/csswg-drafts/issues/824#issuecomment-1204467456)), the Web Animation API is used.

> # Customizing the transition based on the type of navigation

Reviewer: Is there a sample page to point to for a code example? Perhaps something that builds on top of the Navigation API and shows how the 2 integrate nicely.

Author: Wellllll they don't really integrate nicely right now, as you get that "did the navigation work or not??" issue that we're going to fix in the next iteration of the API.

Reviewer: Fair. So we can update this once that issue is fixed in the spec.

In particular, the [Navigation API](https://github.com/WICG/navigation-api) makes it easy to distinguish between a back vs forward traversal/navigation.

# Error handling

This feature is built with the view that a transition is an enhancement to a DOM change. For example:

```js
const transition = new SameDocumentTransition();
await transition.prepare(async () => {
  await updateTheDOMSomehow();
});
```

The API could discover an error before calling the `prepare()` callback, meaning the transition cannot happen. For example, it may discover two elements with the same `page-transition-tag`, or one of the transition elements is fragmented in a way that's incompatible with the API. In this case we still call the `prepare()` callback, because the DOM change is more important than the transition, and being unable to create a transition is not a reason to prevent the DOM change.

However, if a transition cannot be created, the promise returned by `prepare()` will reject.

# Developer tooling

Since this feature is built on existing concepts such as pseudo-elements and CSS animations, tooling for this feature should fit in with existing developer tooling.

> In Chrome's experimental implementation, the pre-existing animation panel could be used to debug transitions, and the pseudo-elements were exposed in the elements panel.

Reviewer: ```suggestion
In Chrome's experimental implementation, the pre-existing animation panel could be used to debug transitions, and the pseudo-elements are exposed in the elements panel.
```

https://user-images.githubusercontent.com/93594/184123157-f4b08032-3b4f-4ca3-8882-8bea0e944355.mp4

# Future work

> There are elements to this feature that we're actively thinking about, but aren't fully designed.

Reviewer: ```suggestion
There are parts to this feature that we're actively thinking about, but aren't fully designed.
```

elements is too overloaded at this point. :P

## Cross-document same-origin transitions

Many developers are more excited about cross-document transitions than SPA transition. We've focused on SPA transitions as it's a slightly smaller problem, and easier to prototype. However, the model has been designed to work across documents.

In the SPA API:

```js
const transition = new SameDocumentTransition();
await transition.prepare(async () => {
  await updateTheDOMSomehow();
});
```

> `prepare()` is used to signal that the current state should be captured, and the promise returned by the promise signals when the new state can be captured, and the transition can begin. Making cross-document transitions work means replacing those moments with moments either side of a cross-document navigation.

Reviewer: ```suggestion
`prepare()` is used to signal that the current state should be captured, and the promise returned by the callback passed to the API signals when the new state can be captured, and the transition can begin. Making cross-document transitions work means replacing those events with implicit events on either side of a cross-document navigation.
```

The rough plan is to allow nesting via an opt-in (all API names used here are for entertainment purposes only):

```css
.container {
  page-transition-tag: container;
  contain: paint;
}
.child-item {
  page-transition-tag: child-item;
  contain: layout;
  page-transition-style-or-whatever: nested;
}
```

With this opt in, rather than the containers being siblings:

```
::page-transition
├─ …
├─ ::page-transition-container(container)
│  └─ ::page-transition-image-wrapper(container)
│     └─ …
└─ ::page-transition-container(child-item)
   └─ ::page-transition-image-wrapper(child-item)
      └─ …
```

…the `child-item` would be nested in its closest parent that's also a transition element:

```
::page-transition
├─ …
└─ ::page-transition-container(container)
   ├─ ::page-transition-image-wrapper(container)
   │  └─ …
   └─ ::page-transition-container(child-item)
      └─ ::page-transition-image-wrapper(child-item)
         └─ …
```

> ## Better pseudo-element selectors

Reviewer: Could we add the content vs full capture point also above this? It has come up as a common use-case when the shape of a UI widget is changing in a transition.

```css
::page-transition-container(foo) {
  & :>> outgoing-image {
    /* … */
  }
  & :>> incoming-image {
    /* … */
  }
}
```

## Transitions targeted to a specific element

In the current design, the transition acts across the whole document. However, developers have expressed interest in using this system, but limited to a single element. For example, allowing two independent components to perform transitions.

This is being discussed in https://github.com/WICG/shared-element-transitions/issues/52.

# Security/Privacy Considerations

The security considerations below cover same-origin transitions.

- Script can never read pixel content in the images. This is necessary since the document may embed cross-origin content (iframes, CORS resources, etc.) and multiple restricted user information (visited links history, dictionary used for spell check, etc.)
- If an element is captured as a 'computed style + content image', any external resources specified on the container, such as background images, will be re-fetched in the context of the new page to account for differences in sandboxing.

Cross-origin transitions aren't yet defined, but are likely to be heavily restricted.

> # Interactivity and accessibility

Reviewer: Not sure if we should add this now since the details need to be sorted in the spec but one more relevant point:

* The duration for which DOM rendering is suppressed, to allow an author to asynchronously switch to the new DOM, input processing is also paused. This is necessary since the visual state presented to the user is inconsistent with the DOM state used for hit-testing.

Reviewer: Sent another PR for this.

> Page transitions are a purely visual affordance. In terms of interactivity, transition elements will behave like `div`s regardless of the original element.

Reviewer: nit: Could these 3 be written as points.

Author: Hmm, but "Developers could break this intent" seems like a follow-on statement, where as a bulleted list is usually for independent statements.

Reviewer: I missed that. So 2 points. :) I put this in a follow up PR.
